xds/internal/xdsclient/client_new.go

```diff
@@ -114,56 +110,42 @@ type OptionsForTesting struct {
 //
 // This function should ONLY be used for testing purposes.
 func NewForTesting(opts OptionsForTesting) (XDSClient, func(), error) {
+	if opts.Name == "" {
+		return nil, nil, fmt.Errorf("opts.Name field must be non-empty")
+	}
 	if opts.WatchExpiryTimeout == 0 {
 		opts.WatchExpiryTimeout = defaultWatchExpiryTimeout
 	}
 	if opts.AuthorityIdleTimeout == 0 {
 		opts.AuthorityIdleTimeout = defaultIdleAuthorityDeleteTimeout
 	}
 
-	// Normalize the input configuration, as this is used as the key in the map
-	// of xDS clients created for testing.
-	buf := bytes.Buffer{}
-	err := json.Indent(&buf, opts.Contents, "", "")
+	cfg, err := bootstrap.NewConfigFromContents(opts.Contents)
 	if err != nil {
-		return nil, nil, fmt.Errorf("xds: error normalizing JSON: %v", err)
+		return nil, nil, err
 	}
-	opts.Contents = bytes.TrimSpace(buf.Bytes())
+	return newRefCountedWithConfig(opts.Name, cfg, opts.WatchExpiryTimeout, opts.AuthorityIdleTimeout)
+}
 
+// GetForTesting returns an xDS client created earlier using the given name.
+//
+// The second return value represents a close function which the caller is
+// expected to invoke once they are done using the client.  It is safe for the
+// caller to invoke this close function multiple times.
+//
+// # Testing Only
+//
+// This function should ONLY be used for testing purposes.
+func GetForTesting(name string) (XDSClient, func(), error) {
```

Comment:
Why did you chose to split this operation up for testing instead of like newRefCountedWithConfig does a get or create in one operation? Is it for ease of use where you create with knobs and then can access by name?

Reply:
Alright that's fair I like that.